provenance/core.py

```diff
@@ -277,13 +315,65 @@ def wrapped(f):
         f._provenance_metadata = t.merge(kargs,
                                          {'version': version,
                                           'name': _name,
+                                          'archive_file': archive_file,
+                                          'delete_original_file': delete_original_file,
                                           'input_hash_fn': input_hash_fn,
                                           'input_process_fn': input_process_fn})
         f.__merge_defaults__ = merge_defaults
         return _provenance_wrapper(repo, f)
 
     return wrapped
 
+class ArchivedFile(object):
+    def __init__(self, id, original_filename=None, in_repo=True):
+        self.blob_id = id
+        self.original_filename = original_filename
+        self.in_repo = in_repo
+
+    def abspath(self):
+        repo = repos.get_default_repo()
+        path = repo.blobstore._filename(self.blob_id)
+        return os.path.abspath(path)
+
+    def __fspath__(self):
+        return self.abspath() if self.in_repo else self.original_filename
+
+    def __str__(self):
+        return self.__fspath__()
+
+    def __repr__(self):
+        return "ArchivedFile<{}>".format(self.blob_id)
+
+
+def file_dump(archived_file, dest_filename, delete_original=False):
+    op = shutil.move if delete_original else shutil.copy
+    op(archived_file.original_filename, dest_filename)
+
+
+def file_load(id):
+    return ArchivedFile(id, in_repo=True)
+
+s.register_serializer('file', file_dump, file_load)
+
+#TODO: allow custom names and custom_fields to be passed in. Will require the provenance wrapper to inspect the function args.
+
+@provenance(archive_file=True, name='archive_file', delete_original_file=True)
+def _archive_file_delete_original(filename, custom_fields=None):
+    return filename
+
+
+@provenance(archive_file=True, name='archive_file', delete_original_file=False)
+def _archive_file_preserve_original(filename, custom_fields=None):
+    return filename
+
+def archive_file(filename, delete_original=False, custom_fields=None):
+    if delete_original:
+        return _archive_file_delete_original(filename, custom_fields)
+    else:
+       return  _archive_file_preserve_original(filename, custom_fields)
```

Comment:
I think the above three methods could be consolidated and accommodate custom fields and names if you just create a provenanced function directly and call it.

```python
def archive_file(filename, delete_original=False, name='archive_file', custom_fields={}):
    @provenance(archive_file=True, name=name, delete_original=delete_original)
    def _archive_file(filename, **custom_fields):
        return filename
    return _archive_file(filename)
```